lib/rust/ensogl/core/src/display/layout/alignment.rs

```diff
@@ -4,110 +4,177 @@ use crate::prelude::*;
+impl Dim1 {
+    pub fn left() -> Self {
+        Self::Start
     }
-
     pub fn center() -> Self {
-        Self::new(Horizontal::Center, Vertical::Center)
+        Self::Center
     }
-    pub fn bottom_left() -> Self {
-        Self::new(Horizontal::Left, Vertical::Bottom)
+    pub fn right() -> Self {
+        Self::End
     }
-    pub fn bottom_right() -> Self {
-        Self::new(Horizontal::Right, Vertical::Bottom)
+    pub fn bottom() -> Self {
+        Self::Start
     }
-    pub fn bottom_center() -> Self {
-        Self::new(Horizontal::Center, Vertical::Bottom)
+    pub fn top() -> Self {
+        Self::End
     }
-    pub fn top_left() -> Self {
-        Self::new(Horizontal::Left, Vertical::Top)
-    }
-    pub fn top_right() -> Self {
-        Self::new(Horizontal::Right, Vertical::Top)
-    }
-    pub fn top_center() -> Self {
-        Self::new(Horizontal::Center, Vertical::Top)
+}
+
+impl Dim1 {
+    pub fn normalized(self) -> f32 {
+        match self {
+            Dim1::Start => 0.0,
+            Dim1::Center => 0.5,
+            Dim1::End => 1.0,
+        }
     }
-    pub fn center_left() -> Self {
-        Self::new(Horizontal::Left, Vertical::Center)
+}
+
+impl Default for Dim1 {
+    fn default() -> Self {
+        Self::Start
     }
-    pub fn center_right() -> Self {
-        Self::new(Horizontal::Right, Vertical::Center)
+}
+
+impl From<&Dim1> for Dim1 {
+    fn from(t: &Dim1) -> Self {
+        *t
     }
 }
 
 
-// === Defaults ===
 
-impl Default for Horizontal {
-    fn default() -> Self {
-        Self::Left
-    }
+// ===============================
+// === 2-dimensional Alignment ===
+// ===============================
+
+#[derive(Clone, Copy, Debug, Deref, Default, PartialEq)]
+pub struct Dim2 {
+    pub vector: Vector2<Dim1>,
 }
-impl Default for Vertical {
-    fn default() -> Self {
-        Self::Bottom
+
+#[allow(missing_docs)]
+impl Dim2 {
+    /// Constructor.
+    pub fn new(horizontal: Dim1, vertical: Dim1) -> Self {
+        Self { vector: Vector2(horizontal, vertical) }
     }
 }
-impl Default for Alignment {
-    fn default() -> Self {
-        let horizontal = default();
-        let vertical = default();
-        Self { horizontal, vertical }
+
+impl Dim2 {
+    pub fn normalized(self) -> Vector2<f32> {
+        Vector2(self.vector.x.normalized(), self.vector.y.normalized())
     }
 }
 
-impl From<&Horizontal> for Horizontal {
-    fn from(t: &Horizontal) -> Self {
-        *t
-    }
+
+
+// ======================================
+// === 2-dimensional Alignment Macros ===
+// ======================================
+
+/// Runs the provided macro with two alignment anchor arrays, one for horizontal and one for
+/// vertical. For example, if run with the arguments `f [args]`, it results in:
+///
+/// ```text
+/// f!{ [args] [left center right] [bottom center top] }
+/// ```
+///
+/// The `[args]` argument is optional.
+#[macro_export]
+macro_rules! with_alignment_dim2_anchors {
```

Comment:
For me, if you have a 2D alignment, it can be represented by 3x3 grid (left-top, center-top, ...) Each of these cells is an "anchor". I like this name, but I can change it if you like something better.